pkg/activator/throttler_test.go

```diff
@@ -0,0 +1,249 @@
+			wantError:       fmt.Sprintf("no revision was found for: %s", revID),
+		},
+	}
+	for _, s := range samples {
+		t.Run(s.label, func(t *testing.T) {
+			var received string
+			want := s.want
+			throttler := getThrottler(s.maxConcurrency, s.revisionGetter, s.endpointsGetter, t)
+			err := throttler.UpdateCapacity(revID, 1)
+			if s.wantError != "" {
+				received = err.Error()
+			}
+			if received != s.wantError {
+				t.Errorf("Expected error in Update capacity. Want %s, got %s", s.wantError, err.Error())
+			}
+			if want > 0 {
+				breaker, _ := throttler.breakers[revID]
+				got := breaker.Capacity()
+				if got != want {
+					t.Errorf("Unexpected capacity of the breaker. Want %d, got %d", want, got)
+				}
+			}
+		})
+	}
+}
+
+func TestThrottler_Try(t *testing.T) {
+	samples := []struct {
+		label           string
+		addCapacity     bool
+		wantBreakers    int32
+		wantError       string
+		revisionGetter  func(RevisionID) (*v1alpha12.Revision, bool, error)
+		endpointsGetter func(RevisionID) (int32, error)
+	}{
+		{
+			label:           "all good",
+			addCapacity:     true,
+			wantBreakers:    int32(1),
+			wantError:       "",
+			revisionGetter:  existingRevisionGetter(10),
+			endpointsGetter: existingEndpointsGetter,
+		},
+		{
+			label:           "non-existing revision",
+			addCapacity:     true,
+			wantBreakers:    int32(0),
+			wantError:       fmt.Sprintf("no revision was found for: %s", revID),
+			revisionGetter:  nonExistingRevisionGetter,
+			endpointsGetter: existingEndpointsGetter,
+		},
+		{
+			label:           "non-existing endpoint",
+			addCapacity:     false,
+			wantBreakers:    int32(0),
+			wantError:       sampleError,
+			revisionGetter:  existingRevisionGetter(10),
+			endpointsGetter: nonExistingEndpointsGetter,
+		},
+	}
+	for _, s := range samples {
+		t.Run(s.label, func(t *testing.T) {
+			var got int32
+			var received string
+			want := s.wantBreakers
+			throttler := getThrottler(defaultMaxConcurrency, s.revisionGetter, s.endpointsGetter, t)
+			if s.addCapacity {
+				throttler.UpdateCapacity(revID, 1)
+			}
+			err := throttler.Try(revID, func() {
+				got++
+			})
+			if s.wantError != "" {
+				received = err.Error()
+			}
+			if received != s.wantError {
+				t.Errorf("Expected error in the Try. Want %s, got %s", s.wantError, received)
+			}
+			if got != want {
+				t.Errorf("Unexpected number of function runs in Try. Want %d, got %d", want, got)
+			}
+		})
+	}
+}
+
+func TestThrottler_Remove(t *testing.T) {
+	throttler := getThrottler(defaultMaxConcurrency, existingRevisionGetter(10), existingEndpointsGetter, t)
+	throttler.breakers[revID] = queue.NewBreaker(throttler.breakerParams)
+	got := len(throttler.breakers)
+	if got != 1 {
+		t.Errorf("Unexpected number of Breakers was created. Want %d, got %d", 1, got)
+	}
+	throttler.Remove(revID)
+	got = len(throttler.breakers)
+	if got != 0 {
+		t.Errorf("Unexpected number of Breakers was created. Want %d, got %d", 0, got)
+	}
```

Comment:
Should this also check that `throttler.breakers[revID]` no longer exists?

Reply:
Isn't it that the len = 0 provides that guarantee?

Reply:
Kinda, but behaviorwise your delete method could also just kill the whole map? The test would still pass.

Reply:
I don't have a strong feeling about that one, could add it. But checking the breaker is not in the map doesn't prevent us from killing the whole map :)